partiql-spi/src/main/java/org/partiql/spi/function/FnProvider.java

```diff
@@ -0,0 +1,214 @@
+package org.partiql.spi.function;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.partiql.spi.types.PType;
+import org.partiql.spi.value.Datum;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * <p>
+ * This provides potentially overloaded implementations of a particular scalar function (e.g. {@code ABS(int)},
+ * {@code ABS(decimal(p, s))}, etc.).
+ * </p>
+ * @see Builder
+ * @see Fn
+ */
+public abstract class FnProvider {
```

Comment:
naming-wise, I think `...Provider` is somewhat non-descriptive. Perhaps a more apt name is `...Overload` since that's how the javadoc describes the `FnProvider` and `AggProvider`?